MetaLamp/lending-pool/src/Plutus/Contracts/AToken.hs

```diff
@@ -0,0 +1,112 @@
+validator _ _ = True
+
+makeLiquidityPolicy :: AssetClass -> MonetaryPolicy
+makeLiquidityPolicy asset = Scripts.mkMonetaryPolicyScript $
+  $$(PlutusTx.compile [|| Scripts.wrapMonetaryPolicy . validator ||])
+    `PlutusTx.applyCode`
+        PlutusTx.liftCode asset
+
+makeAToken :: AssetClass -> AssetClass
+makeAToken asset = assetClass (scriptCurrencySymbol . makeLiquidityPolicy $ asset) (TokenName aTokenName)
+  where
+    (_, tokenName) = unAssetClass asset
+    aTokenName = "a" <> unTokenName tokenName
+
+forgeATokensFrom :: (HasBlockchainActions s) => Aave -> Reserve -> PubKeyHash -> Integer -> Contract w s Text ()
+forgeATokensFrom aave reserve pkh amount = do
+    let script = Core.aaveInstance aave
+        policy = makeLiquidityPolicy (rCurrency reserve)
+        lookups = Constraints.scriptInstanceLookups script
+            <> Constraints.monetaryPolicy policy
+            <> Constraints.ownPubKeyHash pkh
+        aTokenAmount = amount -- / rLiquidityIndex reserve -- TODO: how should we divide?
+        outValue = assetClassValue (rAToken reserve) aTokenAmount
+        tx = mustForgeValue outValue <> mustPayToPubKey pkh outValue
+    ledgerTx <- submitTxConstraintsWith lookups tx
+    _ <- awaitTxConfirmed $ txId ledgerTx
+    pure ()
+
+burnATokensFrom :: (HasBlockchainActions s) => Aave -> Reserve -> PubKeyHash -> Integer -> Contract w s Text ()
+burnATokensFrom aave reserve pkh amount = do
+    let asset = rCurrency reserve
+        aTokenAmount = amount
+        script = Core.aaveInstance aave
+        policy = makeLiquidityPolicy asset
+        burnLookups = Constraints.scriptInstanceLookups script
+            <> Constraints.otherScript (Core.aaveValidator aave)
+            <> Constraints.ownPubKeyHash pkh
+            <> Constraints.monetaryPolicy policy
+        outValue = negate $ assetClassValue (rAToken reserve) aTokenAmount
+        burnTx = mustForgeValue outValue
+    ledgerTx <- submitTxConstraintsWith burnLookups burnTx
+    _ <- awaitTxConfirmed $ txId ledgerTx
+
+    utxos <-
+        Map.filter ((> 0) . flip assetClassValueOf asset . txOutValue . txOutTxOut)
+        <$> utxoAt (Core.aaveAddress aave)
+    let balance = mconcat . fmap (txOutValue . txOutTxOut) . map snd . Map.toList $ utxos
+        remainder = assetClassValueOf balance asset - aTokenAmount
+        withdrawLookups = Constraints.scriptInstanceLookups script
+            <> Constraints.otherScript (Core.aaveValidator aave)
+            <> Constraints.unspentOutputs utxos
+            <> Constraints.ownPubKeyHash pkh
+        orefs = fst <$> Map.toList utxos
+        spendTx = mconcat $ fmap (\ref -> mustSpendScriptOutput ref $ Redeemer $ PlutusTx.toData Core.WithdrawRedeemer) orefs
+        withdrawTx = mustPayToPubKey pkh (assetClassValue asset aTokenAmount)
+            <> spendTx
+            <> mustPayToTheScript Core.DepositDatum (assetClassValue asset remainder)
+    ledgerTx <- submitTxConstraintsWith withdrawLookups withdrawTx
```

Comment:
I think `burnTx` and `withdrawTx` should happen at the same time, otherwise there might be some inconsistencies. E.g. if aTokens are burned and the equivalent Token amount is not sent to the user.